### ISSUE TITLE
asyncssh.SSHClientConnection.run should be able to write stdout to coroutine

### ISSUE THREAD

**Issue by evanfoster** [ronf/asyncssh#139]
Hey there!

I'm using `asyncssh` to take packet traces at remote machines and stream the data back to the client system. It's been working wonderfully, but I've hit the following snag:
```python
connection: asyncssh.SSHClientConnection
async with aiofiles.open(capture_file, mode='wb') as cap_fd, aiofiles.open(log_file, mode='wb') as log_fd:
    async with asyncssh.connect(
            machine, port, client_keys=client_keys, username=user, known_hosts=None, password=password) as connection:
        await connection.run(command, stdout=cap_fd, stderr=log_fd, encoding=None)
```

I really want to use `aiofiles` to open my capture files on the client system. If I just use regular ol' `open` to open these files, everything works, but file writing blocks the event loop. This normally isn't an issue, but there are times when a capture file can grow at hundreds of megabits a second.

Previously, I was defining a custom `SSHClientSession` and was dumping data from that into a queue. I then had a coroutine reading from that queue and writing to a file using aiofiles. I've been refactoring that code to make it easier to use for teaching purposes, and I was really pleased when I found that I could just use `.run` and pass my files to it. It would be perfect except for this one, small issue.

I'm not sure how easy it would be to implement this (maybe you could make an `_AsyncWriter` class?), but it would be greatly appreciated.

**Comment by evanfoster** [comment]
Ahh, I should read things more closely. I'll try subclassing `SSHWriter` and see where that takes me.

**Comment by evanfoster** [comment]
Here's what I ultimately implemented:
```python
class AsyncWriterSSHClientSession(asyncssh.SSHClientSession):
    def __init__(self, cap_fd, log_fd):

        self.log_fd = log_fd
        self.cap_fd = cap_fd

    def data_received(self, data, datatype):
        if datatype == asyncssh.EXTENDED_DATA_STDERR:
            asyncio.ensure_future(self.log_fd.write(data))
        else:
            asyncio.ensure_future(self.cap_fd.write(data))

    def connection_lost(self, exc):
        if exc:
            asyncio.ensure_future(self.log_fd.write(str(exc).encode()))


async def run_client(user: str, machine: str, command: str, capture_file: Path, keys: List[Path], password: str = None,
                     known_hosts: Path = None):
    log_file = capture_file.with_suffix('.log')

    async with aiofiles.open(str(capture_file), mode='wb') as cap_fd, aiofiles.open(str(log_file), mode='wb') as log_fd:
        def session_factory():
            return AsyncWriterSSHClientSession(cap_fd, log_fd)
```
This works for now, but I really dislike having my session factory inside of my context manager. I don't know if you'll want to close this since there's a workable solution, but it would definitely be nice to be able to pass a coroutine to `.run`.

**Comment by ronf** [comment]
What you've got here looks like it should work, but I may be able to make an alternate "AsyncFileWriter" class that is used instead of the regular FileWriter when it detects that the write method is a coroutine. Is that what you mean by "pass a coroutine to run"? The only thing I worry about here is handling file I/O errors, which looks like it might also be an issue with your solution. You're calling ensure_future() without ever collecting the results of the operation, so you'd never see errors that might be raised.

I'd also want to try and do an AsyncFileReader class if I did this, but I suspect that'd be a bit more involved, since I'd actually have to have a coroutine which waiting for the data to come back before feeding it into the SSH channel. Right now, all of the redirection functions are callback-based, not coroutine-based, so there no coroutines are left running to do the I/O.

An alternative here would be to call something like create_process() or run() to get back streams for stdout & stderr (without doing any redirection) and set up your own coroutines for each which alternately read from the channel and then did an async write to the files. That would let you catch any errors which occurred.

**Comment by evanfoster** [comment]
> Is that what you mean by "pass a coroutine to run"?

Correct. The file-like objects that get returned by `aiofiles` have all of the same methods as regular files, just as coroutines. So in this case:
```python
await connection.run(command, stdout=cap_fd, stderr=log_fd, encoding=None)
```
It would be great if the writer would automatically pick up on the fact that `cap_fd.write` is a coroutine.

>  You're calling ensure_future() without ever collecting the results of the operation, so you'd never see errors that might be raised.

You're right about that, and it does worry me a bit. Since the data that's being collected usually isn't that important (I often do `remotecap -w somefile.cap & sleep 5 && wireshark somefile.cap; rm somefile.cap`) I'm not too concerned about it, but there are cases where it would be absolutely essential that the data streamed back gets written to disk.

> An alternative here would be to call something like create_process() or run() to get back streams for stdout & stderr (without doing any redirection) and set up your own coroutines for each which alternately read from the channel and then did an async write to the files. That would let you catch any errors which occurred.

I'd toyed with that, and maybe I'll still go down that route. I do like how compact everything is at the moment though. It's nice to keep things simple when you're using the code for teaching. If people were to start using my little utility for anything serious then I'd probably ditch the unsafe `ensure_future` calls I'm doing and either set up my own readers.

This file redirection feature is is really awesome (I may be making a PR to accept `pathlib` objects if I can find the time.) I really appreciate you taking the time to respond.

**Comment by ronf** [comment]
> It would be great if the writer would automatically pick up on the fact that cap_fd.write is a coroutine.

Yeah - I'll see if I can come up with a clean way to handle this, and reply here if I have something you can test. Thanks for the pointer to aiofiles - I hadn't run across that before.

> This file redirection feature is is really awesome (I may be making a PR to accept pathlib objects if I can find the time.) I really appreciate you taking the time to respond.

Regarding this, I added support for pathlib objects in SFTP in the recently released AsyncSSH 1.12.2. I'm not 100% certain I covered all of the places that I needed to, but if you run into any issue with them not working, please let me know. My primary focus was on allowing pathlib objects for local source & destination paths, but it should also work to use them for remote paths as long as you use the POSIX-compatible pathlib objects and you have an appropriate path encoding set to handle the conversion of the path from Unicode to bytes put into the SFTP packets.

**Comment by ronf** [comment]
I just committed a first cut at supporting this in the develop branch (85977f6). If you get a chance to try this, I'd love to know if it works for you!

**Comment by evanfoster** [comment]
Ah, didn't see this! I'll give it a go.

**Comment by ronf** [comment]
Terrific - thanks for the test!

If you get a chance to try the pathlib support and run into any issues with that, let me know.

I've got one other thing I'm working on right now (dynamic port forwarding via SOCKS) that I should finish up soon and then I'll probably look at rolling these changes into a new release.

**Comment by ronf** [comment]
Both the aiofiles and pathlib support for SSHProcess redirection discussed here are now available in AsyncSSH 1.13.0. If you have any problems with either, please let me know!
